[4.0] NSKeyedUnarchiver methods should take Data (instead of just NSData)

itaiferber opened the pull request:
**What's in this pull request?**
Cherry-picks #10474 and #10515 to `swift-4.0-branch`.

**Explanation:** We missed a few `NSKeyedUnarchiver` methods during our original renaming for Swift — some of these methods still take `NSData` when they can take `Data`. We can add `Data` variants which bridge to `NSData` to fix this in a backwards-compatible way.
**Scope:** Affects those who want to use the proper top-level unarchiving methods in Swift.
**Radar:** rdar://problem/32177014
**Risk:** Low
**Testing:** Manual verification; this adds a new overload to a method we already have with the proper parameter.

Commit:
NSKeyedUnarchiver methods should take Data

We missed a few NSKeyedUnarchiver methods during our original renaming
for Swift — some of these methods still take NSData when they can take
Data. We can add Data variants which bridge to NSData to fix this in a
backwards-compatible way.

itaiferber commented:
@swift-ci Please test

parkera commented on stdlib/public/SDK/Foundation/NSCoder.swift:
```diff
@@ -153,6 +160,13 @@ extension NSKeyedUnarchiver {
+  @nonobjc
+  @available(swift, introduced: 4)
+  @available(OSX 10.11, iOS 9.0, *)
+  public class func unarchiveTopLevelObjectWithData(_ data: Data) throws -> Any? {
```
Shouldn't this be

`unarchiveTopLevelObject(with data: Data)`?

itaiferber replied:
This is the same thing Philippe mentioned. The names already exist as-is; I'm just adding overloads based on the parameter type. If we want to rename them, that'll have to go through API review, no?

itaiferber commented:
Talked to @parkera in person. We'd prefer to take a different approach with this; going to update with a different PR, then cherry-pick here.

Commit:
Change modified method instead of adding overloads

The unarchiveTopLevelObjectWithData which returns Any? (instead of
AnyObject?) was added in Swift 4 and has only shipped in the betas so
far. Instead of adding new overloads which take different types, we
should just fix this one and call it a day.

itaiferber commented:
We'd prefer to integrate #10515 instead.
